pkg/mutation/mutation_function.go

```diff
@@ -0,0 +1,130 @@
+	return mutate(mutator, obj.Object, nil, 0)
+}
+
+func mutate(m Mutator, current interface{}, previous interface{}, depth int) error {
+	if len(m.Path().Nodes)-1 == depth {
+		return addValue(m, current, previous, depth)
+	}
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		next, ok := current.(map[string]interface{})[pathEntry.(*parser.Object).Reference]
+		if !ok {
+			next = createMissingElement(m, current, previous, depth)
+		}
+		if err := mutate(m, next, current, depth+1); err != nil {
+			return err
+		}
+		return nil
+	case parser.ListNode:
+		elementFound := false
+		listPathEntry := pathEntry.(*parser.List)
+		glob := listPathEntry.Glob
+		key := listPathEntry.KeyField
+		for _, listElement := range current.([]interface{}) {
+			if glob {
+				if err := mutate(m, listElement, current, depth+1); err != nil {
+					return err
+				}
+				elementFound = true
+			} else if elementValue, ok := listElement.(map[string]interface{})[key]; ok {
+				if *listPathEntry.KeyValue == elementValue {
+					if err := mutate(m, listElement, current, depth+1); err != nil {
+						return err
+					}
+					elementFound = true
+				}
+			}
+		}
+		// If no matching element in the array was found in non Globbed list, create a new element
+		if !listPathEntry.Glob && !elementFound {
+			next := createMissingElement(m, current, previous, depth)
+			if err := mutate(m, next, current, depth+1); err != nil {
+				return err
+			}
+		}
+	default:
+		return fmt.Errorf("Unrecognized type: %v", t)
+	}
+	return nil
+}
+
+func addValue(m Mutator, current interface{}, previous interface{}, depth int) error {
+	// TODO: it should be considered if the value set can be not just a simple string, but json which could be unmarshalled into an object
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+
+		switch m.(type) {
+		case *AssignMetadataMutator:
+			if elementValue, ok := current.(map[string]interface{})[pathEntry.(*parser.Object).Reference]; ok {
+				log.Info("Mutated value already present", "field", pathEntry.(*parser.Object).Reference, "value", elementValue)
+				return nil
+			}
+		}
+		value, err := m.Value()
+		if err != nil {
+			return err
+		}
+		current.(map[string]interface{})[pathEntry.(*parser.Object).Reference] = value
+	case parser.ListNode:
+		return addListElementWithValue(m, current, previous, depth)
+	}
+	return nil
+}
+
+func addListElementWithValue(m Mutator, current interface{}, previous interface{}, depth int) error {
+	listPathEntry := m.Path().Nodes[depth].(*parser.List)
+
+	if listPathEntry.Glob {
+		return fmt.Errorf("Last path entry can not be globbed")
+	}
+	key := listPathEntry.KeyField
+	keyValue := listPathEntry.KeyValue
+
+	for _, listElement := range current.([]interface{}) {
+		if elementValue, ok := listElement.(map[string]interface{})[key]; ok && keyValue == elementValue {
```

Comment:
use unstructured helpers here.

As written, this may cause a panic if the typecast fails

Reply:
Ustructured functions requires map[string]interface{} as parameter, so we need to cast it anyway.
To be safe, we should wrap it in our own function which checks if the type is correct before casting.

Reply:
+1 some type of typecast. I had forgotten that unstructured helpers didn't deal with this specific issue

Reply:
Added: setNestedField, nestedString